File: packages/core/src/task/ondemand-topology.ts
/**
 * Stage 5 Topology + typed Dependency Graph + On-demand activation
 * (P28 Topology, P29 On-demand, P30 Stage 5, P18 Prefer native parallel).
 *
 * Pure helpers persisted through existing create extras / patch. Does not
 * write Kernel store or task.json. Field names here are an implementation
 * choice — not a frozen Manifest schema. Does not schedule workers.
 */

import { KernelError } from "./kernel-contract.js";
import { isPlainObject } from "./schema.js";

export const STAGE5_SOURCE = "stage5-ondemand-topology";
export const STAGE5_SCHEMA_VERSION = 1 as const;

export const TOPOLOGY_KINDS = ["single", "parent-child"] as const;
export type TopologyKind = (typeof TOPOLOGY_KINDS)[number];

export const DEPENDENCY_EDGE_TYPES = ["requires", "advisory"] as const;
export type DependencyEdgeType = (typeof DEPENDENCY_EDGE_TYPES)[number];

export type Stage5CommandPhase = "create" | "start" | "archive" | "patch";

export const LIFECYCLE_SLOTS = [
  "define",
  "approve",
  "execute",
  "verify",
  "close",
] as const;

export const STAGE5_ONDEMAND_MODULES = [
  "define-extended",
  "independent-check",
  "worker-orchestration",
  "parent-child",
  "debug-recovery",
  "session-transfer",
  "spec-learning",
  "vcs-integration",
  "personal-memory",
  "retention-storage",
  "retrieval-extended",
] as const;

export const PROFILE_BASELINE_MODULES = [
  "intake-basic",
  "define-basic",
  "approval-personal",
  "execute-agent",
  "verify-basic",
  "close-basic",
  "context-progressive",
  "observability-local",
] as const;

export type Stage5OndemandModule = (typeof STAGE5_ONDEMAND_MODULES)[number];

export const ONDEMAND_OWNERS: Readonly<Record<string, string>> = {
  "integration-handoff": "parent-child",
  "session-transfer": "session-transfer",
};

export const LITE_DORMANT_ONDEMAND = [
  "parent-child",
  "vcs-integration",
  "personal-memory",
  "retention-storage",
] as const;

export interface TopologyState {
  schema_version: typeof STAGE5_SCHEMA_VERSION;
  kind: TopologyKind;
  parent_id: string | null;
  children: string[];
}

export interface DependencyEdge {
  from: string;
  to: string;
  type: DependencyEdgeType;
}

export interface DependencyGraph {
  schema_version: typeof STAGE5_SCHEMA_VERSION;
  edges: DependencyEdge[];
}

export interface OndemandTrigger {
  module: string;
  at: string;
  reason: string;
}

export interface OndemandModules {
  schema_version: typeof STAGE5_SCHEMA_VERSION;
  source: typeof STAGE5_SOURCE;
  registered: string[];
  active: string[];
  triggers: OndemandTrigger[];
  owners: Record<string, string>;
  degraded: string[];
}

export interface BaselineModules {
  schema_version: typeof STAGE5_SCHEMA_VERSION;
  source: "profile-runtime";
  active: string[];
}

export interface DecomposeProposal {
  parent_id: string;
  slices: string[];
  confirmed: boolean;
  children_created: string[];
}

export interface CaptureCandidate {
  kind: "candidate";
  summary: string;
  task_created: false;
}

export interface TaskMapGraphProjection {
  graph_authority: "kernel-extras";
  topology_kind: TopologyKind;
  parent_id: string | null;
  children: { id: string; depends_on: string[] }[];
}

export interface Stage5RecordHint {
  id?: string;
  parent?: string | null;
  children?: string[];
}

export function topologyKindFromChildren(
  children: readonly string[],
): TopologyKind {
  return uniqueStrings(children).length > 0 ? "parent-child" : "single";
}

export function defaultTopology(record?: Stage5RecordHint): TopologyState {
  const parent =
    typeof record?.parent === "string" && record.parent.trim() !== ""
      ? record.parent
      : null;
  const children = uniqueStrings(record?.children ?? []);
  return {
    schema_version: STAGE5_SCHEMA_VERSION,
    kind: topologyKindFromChildren(children),
    parent_id: parent,
    children,
  };
}

export function defaultOndemandModules(): OndemandModules {
  return {
    schema_version: STAGE5_SCHEMA_VERSION,
    source: STAGE5_SOURCE,
    registered: [...STAGE5_ONDEMAND_MODULES],
    active: [],
    triggers: [],
    owners: { ...ONDEMAND_OWNERS },
    degraded: [],
  };
}

export function defaultBaselineModules(): BaselineModules {
  return {
    schema_version: STAGE5_SCHEMA_VERSION,
    source: "profile-runtime",
    active: [...PROFILE_BASELINE_MODULES],
  };
}

export function defaultDependencyGraph(): DependencyGraph {
  return { schema_version: STAGE5_SCHEMA_VERSION, edges: [] };
}

export function captureCandidate(summary: string): CaptureCandidate {
  return {
    kind: "candidate",
    summary: summary.trim(),
    task_created: false,
  };
}

export function proposeDecompose(
  parentId: string,
  slices: readonly string[],
): DecomposeProposal {
  return {
    parent_id: parentId,
    slices: uniqueStrings(slices),
    confirmed: false,
    children_created: [],
  };
}

export function confirmDecompose(proposal: DecomposeProposal): DecomposeProposal {
  return {
    ...proposal,
    confirmed: true,
    children_created: [...proposal.slices],
  };
}

export function applyRetention(outcome: string): string {
  return outcome;
}

export function assignParent(
  topology: TopologyState,
  parentId: string,
): TopologyState {
  const next = parentId.trim();
  if (!next) {
    throw new KernelError("INVALID_REQUEST", "parent_id must be a non-empty string");
  }
  if (topology.parent_id && topology.parent_id !== next) {
    throw new KernelError(
      "INVALID_REQUEST",
      `single-parent tree: second parent rejected (${topology.parent_id} vs ${next})`,
    );
  }
  return {
    schema_version: STAGE5_SCHEMA_VERSION,
    kind: topologyKindFromChildren(topology.children),
    parent_id: next,
    children: [...topology.children],
  };
}

export function assertIntegrationAuthority(actorRole: string): void {
  if (actorRole === "parent") return;
  throw new KernelError(
    "INVALID_REQUEST",
    "Integration authority stays with Parent; Child cannot integrate-child",
  );
}

export function expandDependsOnToRequires(
  graph: DependencyGraph,
  fromId: string,
  dependsOn: readonly string[],
): DependencyGraph {
  const edges = [...graph.edges];
  for (const dep of uniqueStrings(dependsOn)) {
    if (hasEdge(edges, fromId, dep, "requires")) continue;
    edges.push({ from: fromId, to: dep, type: "requires" });
  }
  return { schema_version: STAGE5_SCHEMA_VERSION, edges };
}

export function unmetRequires(
  graph: DependencyGraph,
  satisfied: readonly string[],
  fromId?: string,
): string[] {
  const done = new Set(satisfied);
  const missing: string[] = [];
  for (const edge of graph.edges) {
    if (edge.type !== "requires") continue;
    if (fromId && edge.from !== fromId) continue;
    if (done.has(edge.to)) continue;
    if (!missing.includes(edge.to)) missing.push(edge.to);
  }
  return missing;
}

export function projectTaskMapGraph(
  topology: TopologyState,
  graph: DependencyGraph,
): TaskMapGraphProjection {
  const children = topology.children.map((id) => ({
    id,
    depends_on: graph.edges
      .filter((edge) => edge.type === "requires" && edge.from === id)
      .map((edge) => edge.to),
  }));
  return {
    graph_authority: "kernel-extras",
    topology_kind: topology.kind,
    parent_id: topology.parent_id,
    children,
  };
}

export function renderTaskMapProjection(projection: TaskMapGraphProjection): string {
  const lines = [
    "graph_authority: kernel-extras",
    `topology_kind: ${projection.topology_kind}`,
    `parent_id: ${projection.parent_id ?? "null"}`,
    "children:",
  ];
  if (projection.children.length === 0) {
    lines.push("  []");
  } else {
    for (const child of projection.children) {
      const deps =
        child.depends_on.length === 0
          ? "[]"
          : `[${child.depends_on.join(", ")}]`;
      lines.push(`  - id: ${child.id}`);
      lines.push(`    depends_on: ${deps}`);
    }
  }
  return `${lines.join("\n")}\n`;
}

export function residentOnDemandModules(
  extras: Record<string, unknown>,
): string[] {
  return readOndemandModules(extras).active;
}

export function activateOnDemand(
  extras: Record<string, unknown>,
  moduleName: string,
  reason: string,
  at = new Date().toISOString(),
): OndemandModules {
  const current = readOndemandModules(extras);
  if (!current.registered.includes(moduleName)) {
    throw new KernelError(
      "INVALID_REQUEST",
      `On-demand module is not registered: ${moduleName}`,
    );
  }
  const triggers = [
    ...current.triggers,
    { module: moduleName, at, reason: reason.trim() || "trigger" },
  ];
  const active = current.active.includes(moduleName)
    ? [...current.active]
    : [...current.active, moduleName];
  const next: OndemandModules = { ...current, active, triggers };
  extras.ondemand_modules = next;
  return next;
}

export function normalizeTopology(raw: unknown, record?: Stage5RecordHint): TopologyState {
  if (raw === undefined || raw === null) return defaultTopology(record);
  if (!isPlainObject(raw)) {
    throw new KernelError("INVALID_REQUEST", "topology must be a JSON object");
  }
  const kind = raw.kind === "parent-child" ? "parent-child" : raw.kind === "single" ? "single" : null;
  if (kind === null) {
    throw new KernelError(
      "INVALID_REQUEST",
      "topology.kind must be single or parent-child",
    );
  }
  const parentId =
    raw.parent_id === null || raw.parent_id === undefined
      ? null
      : requireId(raw.parent_id, "topology.parent_id");
  const children = uniqueStrings(asStringArray(raw.children));
  const seeded = defaultTopology(record);
  if (parentId && seeded.parent_id && parentId !== seeded.parent_id) {
    throw new KernelError(
      "INVALID_REQUEST",
      `single-parent tree: second parent rejected (${parentId} vs ${seeded.parent_id})`,
    );
  }
  const resolvedParent = parentId ?? seeded.parent_id;
  const resolvedChildren = uniqueStrings([...children, ...seeded.children]);
  return {
    schema_version: STAGE5_SCHEMA_VERSION,
    kind: topologyKindFromChildren(resolvedChildren),
    parent_id: resolvedParent,
    children: resolvedChildren,
  };
}

export function normalizeDependencyGraph(raw: unknown): DependencyGraph {
  if (raw === undefined || raw === null) return defaultDependencyGraph();
  if (!isPlainObject(raw)) {
    throw new KernelError("INVALID_REQUEST", "dependency_graph must be a JSON object");
  }
  const edgesIn = raw.edges;
  if (edgesIn === undefined || edgesIn === null) return defaultDependencyGraph();
  if (!Array.isArray(edgesIn)) {
    throw new KernelError("INVALID_REQUEST", "dependency_graph.edges must be an array");
  }
  const edges: DependencyEdge[] = [];
  for (const item of edgesIn) {
    if (!isPlainObject(item)) {
      throw new KernelError("INVALID_REQUEST", "dependency_graph.edges[] must be objects");
    }
    if (item.type !== "requires" && item.type !== "advisory") {
      throw new KernelError(
        "INVALID_REQUEST",
        "dependency_graph.edges[].type must be requires or advisory",
      );
    }
    const edge: DependencyEdge = {
      from: requireId(item.from, "dependency_graph.edges[].from"),
      to: requireId(item.to, "dependency_graph.edges[].to"),
      type: item.type,
    };
    if (!hasEdge(edges, edge.from, edge.to, edge.type)) edges.push(edge);
  }
  return { schema_version: STAGE5_SCHEMA_VERSION, edges };
}

export function normalizeBaselineModules(raw: unknown): BaselineModules {
  const defaults = defaultBaselineModules();
  if (raw === undefined || raw === null) return defaults;
  if (!isPlainObject(raw)) {
    throw new KernelError("INVALID_REQUEST", "baseline_modules must be a JSON object");
  }
  if (!("active" in raw)) return defaults;
  const allowed = new Set<string>(PROFILE_BASELINE_MODULES);
  return {
    schema_version: STAGE5_SCHEMA_VERSION,
    source: "profile-runtime",
    active: uniqueStrings(asStringArray(raw.active)).filter((name) => allowed.has(name)),
  };
}

export function normalizeOndemandModules(raw: unknown): OndemandModules {
  if (raw === undefined || raw === null) return defaultOndemandModules();
  if (!isPlainObject(raw)) {
    throw new KernelError("INVALID_REQUEST", "ondemand_modules must be a JSON object");
  }
  const defaults = defaultOndemandModules();
  const registered = uniqueStrings([
    ...defaults.registered,
    ...asStringArray(raw.registered),
  ]);
  const active = uniqueStrings(asStringArray(raw.active)).filter((name) =>
    registered.includes(name),
  );
  const degraded = uniqueStrings(asStringArray(raw.degraded));
  const triggers = Array.isArray(raw.triggers)
    ? raw.triggers.filter(isPlainObject).map((item) => ({
        module: typeof item.module === "string" ? item.module : "",
        at: typeof item.at === "string" ? item.at : "",
        reason: typeof item.reason === "string" ? item.reason : "",
      }))
    : [];
  const owners = isPlainObject(raw.owners)
    ? {
        ...defaults.owners,
        ...Object.fromEntries(
          Object.entries(raw.owners).filter(
            (entry): entry is [string, string] => typeof entry[1] === "string",
          ),
        ),
      }
    : { ...defaults.owners };
  return {
    schema_version: STAGE5_SCHEMA_VERSION,
    source: STAGE5_SOURCE,
    registered,
    active,
    triggers,
    owners,
    degraded,
  };
}

export function readTopology(extras: Record<string, unknown>): TopologyState {
  return normalizeTopology(extras.topology);
}

export function readDependencyGraph(extras: Record<string, unknown>): DependencyGraph {
  return normalizeDependencyGraph(extras.dependency_graph);
}

export function readOndemandModules(extras: Record<string, unknown>): OndemandModules {
  return normalizeOndemandModules(extras.ondemand_modules);
}

export function normalizeStage5InExtras(
  extras: Record<string, unknown>,
  record?: Stage5RecordHint,
): void {
  assertUnconfirmedChildCreation(extras, existingTopologyChildren(extras), record);
  extras.topology = normalizeTopology(extras.topology, record);
  let graph = normalizeDependencyGraph(extras.dependency_graph);
  const selfId = typeof record?.id === "string" ? record.id : "";
  const dependsOn = asStringArray(extras.depends_on);
  if (selfId && dependsOn.length > 0) {
    graph = expandDependsOnToRequires(graph, selfId, dependsOn);
  }
  extras.dependency_graph = graph;
  extras.ondemand_modules = normalizeOndemandModules(extras.ondemand_modules);
  extras.baseline_modules = normalizeBaselineModules(extras.baseline_modules);

  const missing = uniqueStrings(asStringArray(extras.ondemand_required_missing));
  if (missing.length > 0) {
    const modules = extras.ondemand_modules as OndemandModules;
    extras.ondemand_modules = {
      ...modules,
      degraded: uniqueStrings([...modules.degraded, ...missing]),
    };
    extras.profile_health = "degraded";
  }

  const topology = extras.topology as TopologyState;
  if (
    topology.kind === "parent-child" &&
    !residentOnDemandModules(extras).includes("parent-child")
  ) {
    activateOnDemand(extras, "parent-child", "topology:parent-child");
  }
}

export function assertStage5ForPhase(
  extras: Record<string, unknown>,
  record: Stage5RecordHint,
  phase: Stage5CommandPhase,
): void {
  if (extras.intake_outcome === "capture-candidate") {
    throw new KernelError(
      "INVALID_REQUEST",
      "Capture Candidate must not create a Task",
    );
  }
  if (isPlainObject(extras.capture_candidate) && extras.capture_candidate.task_created === true) {
    throw new KernelError(
      "INVALID_REQUEST",
      "Capture Candidate must not create a Task",
    );
  }

  assertUnconfirmedChildCreation(extras, existingTopologyChildren(extras), record);
  const topology = normalizeTopology(extras.topology, record);
  extras.topology = topology;
  assertSingleParent(topology, record);

  if (extras.integration_action === "integrate-child") {
    const actor =
      extras.integration_actor === "parent" ? "parent" : "child";
    assertIntegrationAuthority(actor);
  }

  if (phase === "create" || phase === "patch") return;

  assertLifecycleSlots(extras);
  const graph = readDependencyGraph(extras);
  const missing = unmetRequires(
    graph,
    asStringArray(extras.dependency_satisfied),
    typeof record.id === "string" ? record.id : undefined,
  );
  if (missing.length > 0) {
    throw new KernelError(
      "INVALID_TRANSITION",
      `requires unmet: ${missing.join(", ")}`,
    );
  }

  if (phase === "archive") {
    const outcome =
      typeof extras.close_outcome === "string" ? extras.close_outcome : "completed";
    extras.close_outcome = applyRetention(outcome);
  }
}

export function normalizeStage5InExtrasAndAssert(
  extras: Record<string, unknown>,
  record: Stage5RecordHint,
  phase: Stage5CommandPhase,
): void {
  normalizeStage5InExtras(extras, record);
  assertStage5ForPhase(extras, record, phase);
}

function assertSingleParent(topology: TopologyState, record: Stage5RecordHint): void {
  const incoming =
    typeof record.parent === "string" && record.parent.trim() !== ""
      ? record.parent
      : null;
  if (topology.parent_id && incoming && topology.parent_id !== incoming) {
    throw new KernelError(
      "INVALID_REQUEST",
      `single-parent tree: second parent rejected (${topology.parent_id} vs ${incoming})`,
    );
  }
}

function existingTopologyChildren(extras: Record<string, unknown>): string[] {
  if (!isPlainObject(extras.topology)) return [];
  return uniqueStrings(asStringArray(extras.topology.children));
}

function assertUnconfirmedChildCreation(
  extras: Record<string, unknown>,
  existingChildren: readonly string[],
  record?: Stage5RecordHint,
): void {
  const proposal = extras.decompose_proposal;
  if (!isPlainObject(proposal) || proposal.confirmed === true) return;
  const slices = uniqueStrings(asStringArray(proposal.slices ?? proposal.children));
  const incoming = uniqueStrings(record?.children ?? []);
  for (const id of incoming) {
    if (slices.includes(id) && !existingChildren.includes(id)) {
      throw new KernelError(
        "INVALID_REQUEST",
        "Decompose must not create Child before confirmation",
      );
    }
  }
}

function assertLifecycleSlots(extras: Record<string, unknown>): void {
  if (extras.lifecycle_slots === undefined || extras.lifecycle_slots === null) {
    return;
  }
  const present = new Set(asStringArray(extras.lifecycle_slots));
  const missing = LIFECYCLE_SLOTS.filter((slot) => !present.has(slot));
  if (missing.length > 0) {
    throw new KernelError(
      "INVALID_REQUEST",
      `lifecycle required slot missing: ${missing.join(", ")}`,
    );
  }
}

function asStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((item): item is string => typeof item === "string");
}

function uniqueStrings(values: readonly unknown[]): string[] {
  const out: string[] = [];
  for (const value of values) {
    if (typeof value !== "string") continue;
    const id = value.trim();
    if (!id || out.includes(id)) continue;
    out.push(id);
  }
  return out;
}

function requireId(value: unknown, field: string): string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new KernelError("INVALID_REQUEST", `${field} must be a non-empty string`);
  }
  return value.trim();
}

function hasEdge(
  edges: readonly DependencyEdge[],
  from: string,
  to: string,
  type: DependencyEdgeType,
): boolean {
  return edges.some(
    (edge) => edge.from === from && edge.to === to && edge.type === type,
  );
}

File: packages/cli/src/templates/pactile/scripts/common/ondemand_topology.py
"""Stage 5 Topology + typed graph + On-demand (Python projection).

Kernel extras remain authoritative. This module renders and validates the
same contract; it does not schedule workers or write a second graph.
"""

from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

STAGE5_SOURCE = "stage5-ondemand-topology"
STAGE5_SCHEMA_VERSION = 1

# P29 On-demand modules. `integration-handoff` is an owner mapping
# under `parent-child`, not its own activatable module.
STAGE5_ONDEMAND_MODULES = (
    "define-extended",
    "independent-check",
    "worker-orchestration",
    "parent-child",
    "debug-recovery",
    "session-transfer",
    "spec-learning",
    "vcs-integration",
    "personal-memory",
    "retention-storage",
    "retrieval-extended",
)

ONDEMAND_OWNERS = {
    "integration-handoff": "parent-child",
    "session-transfer": "session-transfer",
}

PROFILE_BASELINE_MODULES = (
    "intake-basic",
    "define-basic",
    "approval-personal",
    "execute-agent",
    "verify-basic",
    "close-basic",
    "context-progressive",
    "observability-local",
)

# Documented first-legal-trigger reasons. Tests may pass these; not a hard enum.
ONDEMAND_TRIGGER_HINTS = {
    "worker-orchestration": (
        "isolatable-group-dispatch-window",
        "execution-mode-worker",
        "user-dispatch",
        "independent-check-true-independent",
        "define-extended-research-worker",
    ),
    "retrieval-extended": (
        "collected-evidence-ranking",
        "define-extended-pack",
        "maintainer-probe",
    ),
    "independent-check": (
        "rigor-full",
        "policy-independent-check",
    ),
    "parent-child": (
        "topology-parent-child",
        "user-decompose",
        "multi-deliverable-split",
    ),
}

LITE_DORMANT_ONDEMAND = (
    "parent-child",
    "vcs-integration",
    "personal-memory",
    "retention-storage",
)


class TopologyError(RuntimeError):
    """Topology / On-demand contract rejected a request."""


def capture_candidate(summary: str) -> dict[str, Any]:
    return {
        "kind": "candidate",
        "summary": (summary or "").strip(),
        "task_created": False,
    }


def propose_decompose(parent_id: str, slices: list[str]) -> dict[str, Any]:
    return {
        "parent_id": parent_id,
        "slices": _unique(slices),
        "confirmed": False,
        "children_created": [],
    }


def confirm_decompose(proposal: dict[str, Any]) -> dict[str, Any]:
    slices = _unique(proposal.get("slices") or [])
    return {
        **proposal,
        "confirmed": True,
        "children_created": slices,
    }


def apply_retention(outcome: str) -> str:
    return outcome


def default_topology(
    parent: str | None = None,
    children: list[str] | None = None,
) -> dict[str, Any]:
    """Seed Topology. ``parent_id`` + empty children stays ``single``."""
    kids = _unique(children or [])
    parent_id = (parent or "").strip() or None
    return {
        "schema_version": STAGE5_SCHEMA_VERSION,
        "kind": "parent-child" if kids else "single",
        "parent_id": parent_id,
        "children": kids,
    }


def assign_parent(topology: dict[str, Any], parent_id: str) -> dict[str, Any]:
    current = topology.get("parent_id")
    next_parent = (parent_id or "").strip()
    if not next_parent:
        raise TopologyError("parent_id must be a non-empty string")
    if current and current != next_parent:
        raise TopologyError(
            f"single-parent tree: second parent rejected ({current} vs {next_parent})"
        )
    children = list(topology.get("children") or [])
    return {
        "schema_version": STAGE5_SCHEMA_VERSION,
        "kind": "parent-child" if children else "single",
        "parent_id": next_parent,
        "children": children,
    }


def assert_integration_authority(actor_role: str) -> None:
    if actor_role == "parent":
        return
    raise TopologyError(
        "Integration authority stays with Parent; Child cannot integrate-child"
    )


def expand_depends_on_to_requires(
    graph: dict[str, Any],
    from_id: str,
    depends_on: list[str],
) -> dict[str, Any]:
    edges = list(graph.get("edges") or [])
    for dep in _unique(depends_on):
        if any(
            edge.get("from") == from_id
            and edge.get("to") == dep
            and edge.get("type") == "requires"
            for edge in edges
            if isinstance(edge, dict)
        ):
            continue
        edges.append({"from": from_id, "to": dep, "type": "requires"})
    return {"schema_version": STAGE5_SCHEMA_VERSION, "edges": edges}


def unmet_requires(
    graph: dict[str, Any],
    satisfied: list[str],
    from_id: str | None = None,
) -> list[str]:
    done = set(satisfied)
    missing: list[str] = []
    for edge in graph.get("edges") or []:
        if not isinstance(edge, dict) or edge.get("type") != "requires":
            continue
        if from_id and edge.get("from") != from_id:
            continue
        target = edge.get("to")
        if isinstance(target, str) and target not in done and target not in missing:
            missing.append(target)
    return missing


def project_task_map_graph(
    topology: dict[str, Any],
    graph: dict[str, Any],
) -> dict[str, Any]:
    children = []
    for child_id in topology.get("children") or []:
        if not isinstance(child_id, str):
            continue
        depends = [
            edge.get("to")
            for edge in graph.get("edges") or []
            if isinstance(edge, dict)
            and edge.get("type") == "requires"
            and edge.get("from") == child_id
        ]
        children.append({"id": child_id, "depends_on": depends})
    return {
        "graph_authority": "kernel-extras",
        "topology_kind": topology.get("kind") or "single",
        "parent_id": topology.get("parent_id"),
        "children": children,
    }


def apply_kernel_graph_projection(data: dict[str, Any], extras: dict[str, Any]) -> None:
    """Stamp Parent task-map frontmatter from Kernel extras. Projection only."""
    topology = extras.get("topology")
    graph = extras.get("dependency_graph")
    if not isinstance(topology, dict) and not isinstance(graph, dict):
        return
    projection = project_task_map_graph(
        topology if isinstance(topology, dict) else {},
        graph if isinstance(graph, dict) else {"edges": []},
    )
    data["graph_authority"] = "kernel-extras"
    data["topology_kind"] = projection["topology_kind"]
    if projection.get("parent_id"):
        data["parent_id"] = projection["parent_id"]
    existing = {
        child.get("id"): child
        for child in data.get("children") or []
        if isinstance(child, dict)
    }
    children = data.setdefault("children", [])
    if not isinstance(children, list):
        data["children"] = []
        children = data["children"]
    for child in projection["children"]:
        current = existing.get(child["id"])
        if current is None:
            current = {"id": child["id"], "state": "open", "depends_on": []}
            children.append(current)
            existing[child["id"]] = current
        current["depends_on"] = child["depends_on"]


def resident_on_demand_modules(extras: dict[str, Any]) -> list[str]:
    modules = extras.get("ondemand_modules")
    if not isinstance(modules, dict):
        return []
    active = modules.get("active") or []
    return [name for name in active if isinstance(name, str)]


def default_ondemand_modules() -> dict[str, Any]:
    return {
        "schema_version": STAGE5_SCHEMA_VERSION,
        "source": STAGE5_SOURCE,
        "registered": list(STAGE5_ONDEMAND_MODULES),
        "active": [],
        "triggers": [],
        "owners": dict(ONDEMAND_OWNERS),
        "degraded": [],
    }


def default_baseline_modules() -> dict[str, Any]:
    return {
        "schema_version": STAGE5_SCHEMA_VERSION,
        "source": "profile-runtime",
        "active": list(PROFILE_BASELINE_MODULES),
    }


def extras_from_task_dir(task_dir: Path | str | None) -> dict[str, Any]:
    """Merge Kernel extras with task.json overlay. Missing task → {}."""
    if task_dir is None:
        return {}
    path = Path(task_dir)
    if not path.is_dir():
        return {}
    kernel = _read_json(path / "kernel.json")
    projection = kernel.get("projection") if isinstance(kernel.get("projection"), dict) else {}
    extras = projection.get("extras") if isinstance(projection.get("extras"), dict) else {}
    merged = dict(extras) if isinstance(extras, dict) else {}
    task_data = _read_json(path / "task.json")
    for key in (
        "ondemand_modules",
        "baseline_modules",
        "required_controls",
        "topology",
        "event_bridge",
    ):
        if key not in merged and isinstance(task_data.get(key), dict):
            merged[key] = task_data[key]
    return merged


def is_ondemand_module_active(task_dir: Path | str | None, module_id: str) -> bool:
    """True only after a legal trigger wrote `ondemand_modules.active`."""
    name = (module_id or "").strip()
    if not name:
        return False
    extras = extras_from_task_dir(task_dir)
    return name in resident_on_demand_modules(extras)


def activate_ondemand_module(
    extras: dict[str, Any],
    module_id: str,
    *,
    reason: str,
    at: str | None = None,
) -> dict[str, Any]:
    """First legal trigger writes `active` + `triggers[]` and syncs subscriptions."""
    name = (module_id or "").strip()
    if not name:
        raise TopologyError("ondemand module id must be a non-empty string")
    owner = ONDEMAND_OWNERS.get(name)
    if name == "integration-handoff" or (
        owner == "parent-child" and name not in STAGE5_ONDEMAND_MODULES
    ):
        raise TopologyError(
            "integration-handoff is an owner mapping under parent-child, "
            "not an activatable module"
        )
    if name not in STAGE5_ONDEMAND_MODULES:
        raise TopologyError(f"unknown on-demand module: {name}")

    block = extras.get("ondemand_modules")
    if not isinstance(block, dict):
        block = default_ondemand_modules()
        extras["ondemand_modules"] = block
    registered = block.setdefault("registered", list(STAGE5_ONDEMAND_MODULES))
    if not isinstance(registered, list):
        registered = list(STAGE5_ONDEMAND_MODULES)
        block["registered"] = registered
    if name not in registered:
        registered.append(name)
    active = block.setdefault("active", [])
    if not isinstance(active, list):
        active = []
        block["active"] = active
    if name not in active:
        active.append(name)
    triggers = block.setdefault("triggers", [])
    if not isinstance(triggers, list):
        triggers = []
        block["triggers"] = triggers
    triggers.append(
        {
            "module": name,
            "reason": (reason or "").strip() or "legal-trigger",
            "at": at or _utc_now(),
        }
    )
    _sync_event_bridge(extras)
    return extras


def deactivate_ondemand_module(extras: dict[str, Any], module_id: str) -> dict[str, Any]:
    name = (module_id or "").strip()
    block = extras.get("ondemand_modules")
    if isinstance(block, dict) and isinstance(block.get("active"), list):
        block["active"] = [item for item in block["active"] if item != name]
    _sync_event_bridge(extras)
    return extras


def set_baseline_modules_active(
    extras: dict[str, Any],
    active: list[str],
) -> dict[str, Any]:
    extras["baseline_modules"] = {
        "schema_version": STAGE5_SCHEMA_VERSION,
        "source": "profile-runtime",
        "active": _unique([item for item in active if item in PROFILE_BASELINE_MODULES]),
    }
    _sync_event_bridge(extras)
    return extras


def _sync_event_bridge(extras: dict[str, Any]) -> None:
    try:
        from .adapter_middleware import sync_event_bridge_subscriptions
    except Exception:
        return
    sync_event_bridge_subscriptions(extras)


def _utc_now() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace(
        "+00:00", "Z"
    )


def _read_json(path: Path) -> dict[str, Any]:
    if not path.is_file():
        return {}
    try:
        parsed = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, UnicodeDecodeError, json.JSONDecodeError):
        return {}
    return parsed if isinstance(parsed, dict) else {}


def _unique(values: list[Any]) -> list[str]:
    out: list[str] = []
    for value in values:
        if not isinstance(value, str):
            continue
        item = value.strip()
        if item and item not in out:
            out.append(item)
    return out
